### PR TITLE
[IOPLT-570] Add update activation job

### PR DESCRIPTION
**Pull request by datalek**
#### List of Changes

- Define activation job update endpoint
- Add activation job update handler
- Add activation job update to `ActivationJobWriter`
- Add activation job update implementation
- Replace `NonNegativeNumber` with `NonNegativeInteger`

#### Motivation and Context

The system must allow increasing the number of active users for a trial

#### How Has This Been Tested?

Unit Test

#### Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] I have committed only changes relevant to the task.
- [x] I have minimized the number of changes.
- [ ] My changes are about only one task or issue.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: apps/functions-subscription/src/adapters/azure/functions/__tests__/data.ts [hunk 1]
```diff
@@ -54,3 +54,17 @@ export const makeAValidCreateActivationJobRequest = () =>
       trialId: aTrialId,
     },
   });
+
+export const makeAValidUpdateActivationJobRequest = () =>
+  new HttpRequest({
+    url: 'https://function/trials/{trialId}/activation-jobs/{id}',
```
Comment (kin0992) [inline apps/functions-subscription/src/adapters/azure/functions/__tests__/data.ts:60]: **issue**: wrong path according to the OpenAPI
```suggestion
    url: 'https://function/trials/{trialId}/activation-job,
```

File: apps/functions-subscription/src/adapters/azure/functions/__tests__/update-activation-job.test.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    expect(actual.status).toStrictEqual(200);
+    expect(env.updateActivationJob).toBeCalledWith(trialId, {
+      usersToActivate,
+    });
+  });
+
+  it('should return 400 when the request body is not valid', async () => {
+    const env = makeTestSystemEnv();
+    const aRequestWithInvalidBody = new HttpRequest({
+      url: 'https://function/trials/{trialId}/activation-jobs/{id}',
```
Comment (kin0992) [inline apps/functions-subscription/src/adapters/azure/functions/__tests__/update-activation-job.test.ts:30]: same as above
```suggestion
      url: 'https://function/trials/{trialId}/activation-job',
```

File: apps/functions-subscription/src/adapters/azure/functions/__tests__/update-activation-job.test.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      method: 'PUT',
+      body: { string: '{}' },
+      params: {
+        id: 'anId',
```
Comment (kin0992) [inline apps/functions-subscription/src/adapters/azure/functions/__tests__/update-activation-job.test.ts:34]:
```suggestion
```

File: apps/functions-subscription/src/adapters/azure/functions/__tests__/update-activation-job.test.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      aRequestWithInvalidBody,
+      makeFunctionContext(),
+    );
+
+    expect(actual.status).toStrictEqual(400);
+    expect(await actual.json()).toMatchObject({
+      status: 400,
+      detail: 'Missing or invalid body',
+    });
+    expect(env.insertActivationJob).toBeCalledTimes(0);
```
Comment (kin0992) [inline apps/functions-subscription/src/adapters/azure/functions/__tests__/update-activation-job.test.ts:49]:
```suggestion
    expect(env.updateActivationJob).toBeCalledTimes(0);
```

File: apps/functions-subscription/src/adapters/azure/functions/update-activation-job.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      'requestBody',
+      RTE.fromEither(parseRequestBody(UpdateActivationJob)(req)),
+    ),
+    RTE.flatMapTaskEither(({ updateActivationJob, trialId, requestBody }) =>
+      updateActivationJob(trialId, {
+        usersToActivate: requestBody.usersToActivate as NonNegativeInteger,
+      }),
+    ),
+    RTE.map(flow(toActivationJobAPI, H.successJson, H.withStatusCode(200))),
+    RTE.mapLeft(toHttpProblemJson),
```
Comment (kin0992) [inline apps/functions-subscription/src/adapters/azure/functions/update-activation-job.ts:36]: **note**: I think you can do this as well
```suggestion
    RTE.mapBoth(toHttpProblemJson, flow(toActivationJobAPI, H.successJson)),
```

**Review by kin0992 — CHANGES_REQUESTED** [review]

File: apps/functions-subscription/api/internal.yaml [hunk 6]
```diff
@@ -100,6 +100,38 @@ paths:
...
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ActivationJob'
+        '400':
+          description: Bad Request
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ProblemJson'
```
[the remark this answers is not in the document]
Comment (datalek) [reply apps/functions-subscription/api/internal.yaml]: Done ✅

File: apps/functions-subscription/src/adapters/azure/cosmosdb/activation-job.ts [hunk 7]
```diff
@@ -29,5 +29,25 @@ export const makeActivationJobCosmosContainer = (
...
       ),
+    update: (trialId, update) =>
+      pipe(
+        TE.tryCatch(
+          () =>
+            container.item(trialId, trialId).patch([
+              {
+                op: 'replace',
+                path: '/usersToActivate',
+                value: update.usersToActivate,
```
Comment (kin0992) [inline apps/functions-subscription/src/adapters/azure/cosmosdb/activation-job.ts:40]: **issue**: `usersToActivate` is defined as an optional parameter; what happens if that parameter is `undefined`?

**Pull request merged**